Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs

```diff
@@ -119,31 +118,37 @@ public void Delete(IEnumerable<ParticipantScore> participantScoresEnumerable)
             this.participantScores.SaveChanges();
         }
 
-        public void AddBySubmissionByUsernameAndIsOfficial(Submission submission, string username, bool isOfficial)
+        public void AddBySubmissionByUsernameAndIsOfficial(
+            Submission submission,
+            string userName,
+            Participant participant)
         {
-            this.participantScores.Add(new ParticipantScore
+            participant.Scores.Add(new ParticipantScore
             {
                 ParticipantId = submission.ParticipantId.Value,
                 ProblemId = submission.ProblemId.Value,
                 SubmissionId = submission.Id,
-                ParticipantName = username,
+                ParticipantName = userName,
                 Points = submission.Points,
-                IsOfficial = isOfficial
+                IsOfficial = participant.IsOfficial
             });
+            participant.TotalScoreSnapshot += submission.Points;
 
-            this.participantScores.SaveChanges();
+            this.participantsData.Update(participant);
         }
 
         public void UpdateBySubmissionAndPoints(
             ParticipantScore participantScore,
             int? submissionId,
-            int submissionPoints)
+            int submissionPoints,
+            Participant participant)
         {
+            participant.TotalScoreSnapshot = (participant.TotalScoreSnapshot - participantScore.Points) + submissionPoints;
```

Comment:
Is it guaranteed that `participant.TotalScoreSnapshot` will always be greater or equal to `participantScore.Points`?

Reply:
Yes, this logic is inside UpdateBydSubmissionAndPoints, therefore in worst case the participant.TotalScoreSnapshot will be equal to participantScore.Points. On the other hand the submissionPoints will always be greater than participantScore.points